lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp

```diff
@@ -95,16 +96,120 @@ SILCombiner::visitRefToRawPointerInst(RefToRawPointerInst *rrpi) {
   return nullptr;
 }
 
-SILInstruction *SILCombiner::visitUpcastInst(UpcastInst *UCI) {
-  if (UCI->getFunction()->hasOwnership())
+namespace {
+
+/// A folder object for sequences of forwarding instructions that forward owned
+/// ownership. Is used to detect if we can delete the intermediate forwarding
+/// instructions without ownership issues and then allows the user to either
+/// delete all of the rest of the forwarding instructions and then replace front
+/// with a new value or set front's operand to a new value.
+class SingleBlockOwnedForwardingInstFolder {
+  SmallVector<SingleValueInstruction *, 4> rest;
+  SILCombiner &SC;
+  SingleValueInstruction *front;
+
+public:
+  SingleBlockOwnedForwardingInstFolder(
+      SILCombiner &SC, SingleValueInstruction *instructionToFold)
+      : SC(SC), front(instructionToFold) {
+    // If our initial instruction to fold isn't owned, set it to nullptr to
+    // indicate invalid.
+    if (SILValue(instructionToFold).getOwnershipKind() != OwnershipKind::Owned)
+      front = nullptr;
+  }
+
+  bool isValid() const { return bool(front); }
+
+  bool add(SingleValueInstruction *next) {
+    assert(isValid());
+    if (SILValue(next).getOwnershipKind() != OwnershipKind::Owned)
+      return false;
+
+    if (next->getSingleUse()) {
+      rest.push_back(next);
+      return true;
+    }
+
+    if (front->getParent() != next->getParent()) {
+      return false;
+    }
+
+    // Otherwise, since the two values are in the same block and we want to
+    // optimize only if our original value doesn't have any non-debug uses, we
+    // know that our value can only have a single non-debug use, the consuming
+    // user. So if we are not in that situation, bail.
+    if (!hasOneNonDebugUse(next))
+      return false;
+
+    rest.push_back(next);
+    return true;
+  }
+
+  /// Delete all forwarding uses and then RAUW front with newValue.
+  SingleValueInstruction *optimizeWithReplacement(SILValue newValue) && {
+    // NOTE: Even though after running cleanup rest, front now has its
+    // forwarding operand set to Undef, we haven't touched its result. So it is
+    // safe to RAUW.
+    cleanupRest();
+    SC.replaceValueUsesWith(front, newValue);
```

Comment:
Question: should we also be erasing `front` here since it returns `nullptr`?

Reply:
I did this on purpose since I wanted to signal to the caller/SILCombine not to do anything further. I took care of everything.

Reply:
@gottesmm my point is that if you are taking care of everything (hence returning nullptr) shouldn't you be erasing the dead `front` instruction?